syntax = "proto3";

option csharp_namespace = "Gymize.Protobuf";

// Tensors use the C-like index order,
// with the last axis index changing fastest,
// back to the first axis index changing slowest.

message TensorProto {
	bytes data = 1;
	string dtype = 2;
	repeated int32 shape = 3;
}

message GraphProto {
	TensorProto nodes = 1; // floating type
	TensorProto edges = 2; // floating type
	TensorProto edge_links = 3; // integer type
}

enum CompressionTypeProto {
	COMPRESSION_TYPE_PROTO_UNSPECIFIED = 0;
	COMPRESSION_TYPE_PROTO_NONE = 1;
	COMPRESSION_TYPE_PROTO_PNG = 2;
	COMPRESSION_TYPE_PROTO_JPG = 3;
}

message ImageProto {
	CompressionTypeProto compression_type = 1;
	bytes data = 2;
	string dtype = 3;
	repeated int32 shape = 4; // (H, W, C) or (H, W)
	repeated int32 transpose_axes = 5; // H = 0, W = 1, C = 2
}

enum InstanceTypeProto {
	INSTANCE_TYPE_PROTO_UNSPECIFIED = 0;
	INSTANCE_TYPE_PROTO_RAW = 1;
	INSTANCE_TYPE_PROTO_TENSOR = 2;
	INSTANCE_TYPE_PROTO_DISCRETE = 3;
	INSTANCE_TYPE_PROTO_TEXT = 4;
	INSTANCE_TYPE_PROTO_DICT = 5;
	INSTANCE_TYPE_PROTO_LIST = 6;
	INSTANCE_TYPE_PROTO_GRAPH = 7;
	// The following are gymize-defined types, not in gym
	INSTANCE_TYPE_PROTO_IMAGE = 8;
	INSTANCE_TYPE_PROTO_FLOAT = 9;
	INSTANCE_TYPE_PROTO_BOOL = 10;
	INSTANCE_TYPE_PROTO_NULL = 11;
	INSTANCE_TYPE_PROTO_JSON = 12;
}

message InstanceProto { // Instance is a instance belongs to the space
	InstanceTypeProto type = 1;
	bytes raw_data = 2;
	TensorProto tensor = 3;
	int64 discrete = 4;
	string text = 5;
	map<string, InstanceProto> dict = 6;
	repeated InstanceProto list = 7; // Tuple and Sequence
	GraphProto graph = 8;
	// The following are gymize-defined types, not in gym
	ImageProto image = 9; // Box space instance which is a compressed image (H, W, C)
	double float = 10; // float should turn to array in Box space
	bool boolean = 11; // boolean is kind of discrete
	string json = 12;
}